src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs

```diff
@@ -1307,6 +1307,28 @@ internal static bool IsGenericTaskType(this TypeSymbol type, CSharpCompilation c
             return namedType.IsCustomTaskType(out builderArgument);
         }
 
+        internal static bool IsAsyncMainReturnType(this TypeSymbol type, CSharpCompilation compilation)
+        {
+            var namedType = type as NamedTypeSymbol;
+            if (namedType == null)
+            {
+                return false;
+            }
+            else if (namedType.IsNonGenericTaskType(compilation))
+            {
+                return true;
+            }
+            else if ((object)namedType.ConstructedFrom == compilation.GetWellKnownType(WellKnownType.System_Threading_Tasks_Task_T))
+            {
+                return namedType.TypeParameters.Length == 1 && 
```

Comment:
this check is unnecessary.  If your ```ConstructedFrom == Task<>```, then your TypeParameter.Length is 1 by invariance.